modules/gci.py

```
@@ -0,0 +1,152 @@
+import sqlite3
+import time
+from tools import DatabaseCursor, db_path
+
+def setup(self):
+    self.gci_data = {}
+
+    self.gci_db = db_path(self, 'gci')
+
+    connection = sqlite3.connect(self.gci_db)
+    cursor = connection.cursor()
+
+    cursor.execute('''create table if not exists gci_data (
+        nick        varchar(255),
+        full_name   varchar(255),
+        all_time    unsigned big int not null default 0,
+        unique (nick, full_name) on conflict replace
+    );''')
+
+    cursor.close()
+    connection.close()
+
+def teardown(self):
+    with DatabaseCursor(self.gci_db) as cursor:
+        cursor.execute("SELECT nick FROM gci_data")
+
+        rows = cursor.fetchall()
+
+    for row in rows:
+        nick = row[0]
+        stop(self, nick)
+
+def commutate(phenny, nick):
+    nick = nick.casefold()
+
+    now_time = time.time()
+    last_time = phenny.gci_data[nick]['last_time']
+    new_time = now_time - last_time
+
+    sqlite_data = {
+        'nick': nick,
+    }
+
+    with DatabaseCursor(phenny.gci_db) as cursor:
+        cursor.execute("SELECT full_name, all_time FROM gci_data WHERE nick=:nick", sqlite_data)
+        full_name, all_time = cursor.fetchone()
+
+        all_time += new_time
+        sqlite_data.update({
+            'all_time': all_time,
+        })
+
+        cursor.execute('update gci_data set all_time=:all_time where nick=:nick', sqlite_data)
+
+    phenny.gci_data[nick]["last_time"] = now_time
+
+    if all_time - new_time <= 4*60*60 <= all_time:
+        phenny.msg("#apertium", full_name + " stayed on the IRC channel for four hours.")
+
+def start(phenny, nick):
+    nick = nick.casefold()
+
+    if nick in phenny.gci_data:
```

Such would indicate that the function may have been erroneously called, and so an exception should be thrown instead of an immediate return to the caller. (The caller should validate if the nick is inside whenever necessary.)